Modules/_operator.c

```diff
@@ -1595,36 +1595,59 @@ static PyType_Spec attrgetter_type_spec = {
 typedef struct {
     PyObject_HEAD
     PyObject *name;
-    PyObject *xargs; // reference to arguments passed in constructor
+    PyObject *args;
     PyObject *kwds;
     PyObject **vectorcall_args;  /* Borrowed references */
     PyObject *vectorcall_kwnames;
```

Comment:
The ownership here seems a bit complicated and I think it can be simplified. As I understand it, `vectorcall_kwnames` only exists to ensure that some entries in `vectorcall_args` stay alive.

Instead, I'd suggest:

* Make `PyObject *vectorcall_args` a tuple (that holds strong references to its contents as usual)
* Get rid of `vectorcall_kwnames`
* Use `_PyTuple_ITEMS` for fast access to the contents of the tuple (for `memcpy()`)
* Visit  `vectorcall_args` in `methodcaller_traverse`

Reply:
The `vectorcall_kwnames` is needed as an argument for `PyObject_VectorcallMethod` in `methodcaller_vectorcall` (https://github.com/python/cpython/blob/main/Modules/_operator.c#L1666), so we cannot get rid of it.

The ownership is not too hard I think: the objects in `vectorcall_args` have references borrowed from either `mc->args` or (the keys from) `mc->kwds`. I added a comment to clarify this.

Making the `vectorcall_args` a tuple is still an option though. It requires a bit more memory and a tiny bit of computation in the initialization. It would be the C equivalent of `vectorcall_args = args + tuple(kwds)`. I'll work it out in a branch to see how it compares

Reply:
Okay, in that case don't worry about it unless you prefer it as a tuple.

Reply:
The diff between the two approaches is this:

https://github.com/eendebakpt/cpython/compare/methodcaller_ft...eendebakpt:cpython:methodcaller_ft_v2

What is nice about making `vectorcall_args` a tuple is that if there are no keyword arguments, we can reuse `mc->args`. It does require more operations in the construction though.

Reply:
I think either approach is fine! My guess is that the vast majority of [uses](https://github.com/search?q=%2Fmethodcaller%5C%28%2F&type=code) of `methodcaller()` do not use keyword arguments.

Reply:
Running benchmarks shows what is to be expected from the implementations: using a tuple for `vectorcall_args` is a bit slower in initializing, except when there are no keyword arguments (since then we reuse the `arg` tuple). Differences are small though.

Since using a tuple leads to cleaner code and the majority of uses is without keywords I slightly prefer the tuple approach. I will open a new PR for it.